### PR TITLE
Add server hooks that provide access to request/response

### PR DESCRIPTION
**Pull request by Cyberax**
Current hooks can only inspect metadata, the can't be used for logging
or validation. Add RequestDeserialized and ResponsePrepared hooks that
have access to raw Go request and response objects.

*Issue #, if available:*
https://github.com/twitchtv/twirp/issues/212

*Description of changes:*
Added two more server hooks

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: install_proto.bash [hunk 1]
```diff
@@ -22,7 +22,7 @@ if [ $PROTOC_EXISTS -eq 0 ]; then
 		exit 0
 	fi
 	echo "libprotoc 3.11.0 required, but found: $PROTOC_VERSION"
-	exit 1
+	#exit 1
```
Comment (iheanyi) [inline install_proto.bash]: Should this be reverted?

Comment (Cyberax) [reply install_proto.bash]: Yes, of course. Fixed.

**Comment by Cyberax** [comment]
Any chance you can merge this?

**Comment by dpolansky** [comment]
Hi @Cyberax, thanks for contributing the issue and PR. I've taken a look at the change, but I need additional time to discuss the issue and its motivations further with @spenczar. Thanks for your patience, and my apologies for the slow turnaround time here.

**Comment by Cyberax** [comment]
No worries! Thanks for looking at it!

I'm also thinking about hooks that can modify/decorate the errors. For example, add the RequestId as a metadata field for tracing.

**Pull request closed without merging**